### DISCUSSION TITLE
How to use migration

### DISCUSSION THREAD

**Discussion by meyer0x** [prisma/prisma#12202] [Q&A] [answered]
Hello to all,

I don't understand how migration works and I'm about to put an application in production.

First of all how do you manage the migration folder on GitHub? 

In my prod branch I didn't put it in the .gitignore but my migrations folder is empty. Do I have to keep the migrations especially in the prod branch?

How to do if I deleted my migration folder to keep the data? I have done some staging tests and when I do migrate deploy it asks me to reset the database

If someone could explain me how to use the migrations it would be with pleasure.

_I want to remind you that I have seen the documentation and that is why I ask for help here_

**Comment by iiian** [comment, marked as answer]
So I understand that you've read the documentation, but given that your question here is still pretty broad, I think anybody runs the risk of being condescending so please forgive me if I suggest things you've already considered or that seem obvious.

Any approach to doing migrations requires getting your database schema "up-to-date". _How_ you do migrations ultimately boils down to what your non-functional requirements are for your context and environment. For instance: Do things need to be automated (eg CI/CD driven, gitops)? Do you have multiple databases to migrate? How do you want to handle rollbacks, in case of an issue? These are the types of questions to consider, although you might be able to just get away with the default simple answer: I can do migrations by-hand, I have only one database, if something goes wrong I'm going to wipe the database and start from scratch.

If you're in such a default case, you can really just run `npx prisma migrate deploy` or `npx prisma db push`. The next question is _how to specify the prod database._ If you bake the url into your `datasource`, for example:

```prisma
datasource db {
  provider = "postgresql"
  url      = "postgresql://usr:pas@host:5432/db"
}
``` 

you're going to have a bad time, because there's nothing you can really do here to specify a JIT prod db connections string. To fix this, use `env("DATABASE_URL")`.

Once you do that, you can JIT a db url with the shell command as such:

```bash
DATABASE_URL=postgresql://usr2:pas2@prod.host.com:5432/prod_db npx prisma migrate deploy
```

That's how I would think to do it at least, there are certainly other ways to do it!

**Reply by meyer0x** [reply]
Good evening Iaan,

First of all, thank you for your answer.

Indeed I knew about the env in the schema.prisma.

I think I expressed myself badly and I apologize. When I ask how to do in production I mean that I would like to know the flow for migration.

Actually I think it is this:

I develop on my dev branch a feature.
I merge on prod
in prod branch ( with old migrations ) npx prisma migrate dev

and then with docker npx prisma migrate deploy

but how to do if I deleted the migration folder ? I tried in staging and it asks me to reset the database because I added a table and columns that's why I'm afraid for my future production.

Sorry if I express myself badly, thanks again for your time.

**Reply by iiian** [reply]
Oh and to your question about the migration folder. The migration folder is a "stateful-artifact" produced by `prisma migrate`. It tracks the incremental progression of schema state in each of your databases: dev, test, prod. Note that you don't _need_ to have a dev and test and prod database. You can have just test & prod, or just prod. It's all about how you want to do you SDLC (software development lifecycle). For simplicity's sake, let's just pretend we've got a single "prod" database, which is really where we're doing development, testing it, and finally serving real customers. (You wouldn't want to do this in the real world most of the time).  

In this scenario, `prisma migrate` produces two things: (1) locally, this `migrations` folder, and (2) in the remote database a `_prisma_migrations` table in your database. If you check the `_prisma_migrations` table, you'll see that it creates a record for a migration in the `migration_name` column. If you check your `migrations` folder, you'll see there's a folder by the same name, containing the sql script that was applied to bring the database to that state.

So basically, you may _want_ to keep the `migrations` folder in version control, depending on your context, or you may not want to. If you're going to be bound to just a handful of databases for a long time, it's probably good to keep `migrations` in version control. If you're building a system that might be replicated to lots and lots of other databases, maybe not so much, as the migrations don't necessarily offer much benefit.

**Reply by meyer0x** [reply]
Indeed I can't see myself using only one database which is also the one for production.

Thank you for all these answers, it starts to be a little clearer. 

I will try tomorrow with a development and test database as if it was in production. While keeping the migrations folder because I don't want to lose the data.

Thanks for everything, I'll keep you posted tomorrow ^^

Have a nice evening

**Reply by iiian** [reply]
> Actually I think it is this:
>
> I develop on my dev branch a feature.
> I merge on prod
>in prod branch ( with old migrations ) npx prisma migrate dev

Hmm, maybe this is causing some issues? I don't know if I understand why `npx prisma migrate dev` in the prod branch 🤔 🤷‍♂️ 

>and then with docker npx prisma migrate deploy

Are you saying you like `docker run -it <my-prod-container> npx prisma migrate deploy`? Yeah, I think having the migrations folder would probably help here.

>but how to do if I deleted the migration folder ? 

I think you can regenerate it in the dev environment and push it to prod as a fix

**Reply by iiian** [reply]
> Indeed I can't see myself using only one database which is also the one for production.
> 
> Thank you for all these answers, it starts to be a little clearer.
> 
> I will try tomorrow with a development and test database as if it was in production. While keeping the migrations folder because I don't want to lose the data.
> 
> Thanks for everything, I'll keep you posted tomorrow ^^
> 
> Have a nice evening

Hope this helps! Best of luck.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
